x-pack/plugins/maps/public/shared/layers/sources/wms_source/wms_client.js

```diff
@@ -0,0 +1,129 @@
+
+    const parsePromise = new Promise((resolve, reject) => {
+      parseString(body, (error, result) => {
+        if (error) {
+          reject(error);
+        } else {
+          resolve(result);
+        }
+      });
+    });
+    return await parsePromise;
+  }
+
+  reduceLayers(path, layers) {
+    const emptyCapabilities = {
+      layers: [],
+      styles: [],
+    };
+    function createOption(optionPath, optionTitle, optionName) {
+      return {
+        path: [...optionPath, optionTitle],
+        value: optionName
+      };
+    }
+
+    return layers.reduce((accumulatedCapabilities, layer) => {
+      // Layer is hierarchical, continue traversing
+      if (layer.Layer) {
+        const hierarchicalCapabilities  = this.reduceLayers([...path, layer.Title[0]], layer.Layer);
+        return {
+          layers: [...accumulatedCapabilities.layers, ...hierarchicalCapabilities.layers],
+          styles: [...accumulatedCapabilities.styles, ...hierarchicalCapabilities.styles]
+        };
+      }
+
+      const updatedStyles = [...accumulatedCapabilities.styles];
+      if (_.has(layer, 'Style[0]')) {
+        updatedStyles.push(createOption(
+          path,
+          _.get(layer, 'Style[0].Title[0]'),
+          _.get(layer, 'Style[0].Name[0]')
+        ));
+      }
+      return {
+        layers: [
+          ...accumulatedCapabilities.layers,
+          createOption(path, layer.Title[0], layer.Name[0])
+        ],
+        styles: updatedStyles
+      };
+    }, emptyCapabilities);
+  }
+
+  // Avoid filling select box option label with text that is all the same
+  // Create a single group from common parts of Layer hierarchy
+  groupCapabilities(list) {
+    if (list.length === 0) {
+      return [];
+    }
+
+    let rootCommonPath = list[0].path;
+    for(let listIndex = 1; listIndex < list.length; listIndex++) {
+      if (rootCommonPath.length === 0) {
+        // No commonality in root path, nothing left to verify
+        break;
+      }
+
+      const path = list[listIndex].path;
+      for(let pathIndex = 0; pathIndex < path.length && pathIndex < rootCommonPath.length; pathIndex++) {
+        if (rootCommonPath[pathIndex] !== path[pathIndex]) {
+          // truncate root common path at location of divergence
+          rootCommonPath = rootCommonPath.slice(0, pathIndex);
+          break;
+        }
+      }
+    }
+
+    if (rootCommonPath.length === 0 || list.length === 1) {
+      return list.map(({ path, value }) => {
+        return { label: path.join(' - '), value };
+      });
+    }
+
+    return [{
+      label: rootCommonPath.join(' - '),
+      options: list.map(({ path, value }) => {
+        return { label: path.splice(rootCommonPath.length).join(' - '), value };
+      })
+    }];
+  }
+
+  async getCapabilities() {
+    const rawCapabilities = await this._fetchCapabilities();
+
+    const { layers, styles } = this.reduceLayers([], _.get(rawCapabilities, 'WMT_MS_Capabilities.Capability[0].Layer', []));
```

Comment:
`WMT_MS_Capabilities.Capability[0].Layer` will always work (?)

Can you point to spec that top-level is always `WMT_MS_Capabilities`?

Reply:
WMS spec is available here - https://www.opengeospatial.org/standards/wms.